src/librustc/traits/project.rs

```diff
@@ -818,55 +819,57 @@ fn project_type<'cx, 'gcx, 'tcx>(
                                        &obligation_trait_ref,
                                        &mut candidates);
 
```

Comment:
This core idea here seems great, but -- as you said -- this setup feels a bit awkward to read. I think we can make the code flow much nicer, however. Perhaps we should change the "candidate set" data structure into something like this:

```rust
enum ProjectionTyCandidateSet<'tcx> {
    None,
    Single(ProjectionTyCandidate<'tcx>),
    Ambiguous,
}

impl<'tcx> ProjectionTyCandidateSet<'tcx> {
    fn mark_ambiguous(&mut self) {
        *self = ProjectionTyCandidateSet::Ambiguous;
    }

    // Returns true if the push was successful, or false if the candidate
    // was discarded -- this could be because of ambiguity, or because
    // a higher-priority candidate is already there.
    fn push_candidate(&mut self, candidate: ProjectionTyCandidate<'tcx>) -> bool {
        match self {
            ProjectionTyCandidateSet::None => {
                *self = ProjectionTyCandidateSet::Single(candidate);
                true
            }

            ProjectionTyCandidateSet::Single(ProjectionTyCandidate::ParamEnv(..)) => {
                if let ProjectionTyCandidate::ParamEnv(..) = candidate {
                    *self = ProjectionTyCandidateSet::Ambiguous;
                }
                false
            }

            ProjectionTyCandidateSet::Single(..) => {
                *self = ProjectionTyCandidateSet::Ambiguous;
                false
            }

            ProjectionTyCandidateSet::Ambiguous => {
                false
            }
        }
    }
}
```

Then, each time that we now "push onto the set", we can instead invoke `push_candidate`. This lets us drop the closure -- instead we use the return value of this method to decide whether to "keep" the snapshot.